Handle USD PointInstancer - Rendering and Point Promotion

danieldresser-ie opened the pull request:
I think this in a fairly good place for review. This should let you load a USD file containing a PointInstancer, immediately see it rendering in Arnold, and select points to promote to live geo.

There are lots of names here that I'm not very confident in - happy to switch to whatever wording feels best to people.

The PromoteInstances tool itself is currently a naive adaptation of an ExtensionAlgo.exportExtension - I think for a tool like this, I would probably prefer to reformat it to be a bit more maintainable, and probably blow away the backdrops and uiPositions - but it might be easier for you to inspect interactively in it's current form. There are a couple of places where we could simplify it if we added simple features to the C++ nodes, but I thought it makes sense to look at the current version first.

One thing I'll mention in terms of user experience: you clarified that you don't consider this a bug, but it still feels weird in Cycles when no matter how much I shift click in the Hierarchy View to expand everything, I still just see a bounding box for my instances, and the only way to get them to actually show up is to turn on "Expand All" in the expansion menu. ( It works exactly how you'd expect in Arnold, the instances appearing when you expand the location of the instancer - that's because we can use encapsulation in Arnold ).

johnhaddon commented on src/GafferScene/BranchCreator.cpp:
```diff
@@ -131,6 +131,52 @@ void validateDestination( const ScenePlug::ScenePath &destination )
+	const ScenePlug::ScenePath &queryPath, const ScenePlug::ScenePath &refPath, const ScenePlug *inPlug
+)
+{
+	unsigned int matchingLength = std::min( queryPath.size(), refPath.size() );
+	for( unsigned int i = 0; i < matchingLength; i++ )
+	{
+		if( queryPath[ i ] != refPath[ i ] )
+		{
+			matchingLength = i;
+			break;
+		}
+	}
+
+	IECore::CompoundObjectPtr result;
+
+	ScenePlug::ScenePath curPath = queryPath;
+	curPath.resize( matchingLength + 1 );
+	for( unsigned int i = matchingLength; i < queryPath.size(); i++ )
```
I've got a feeling this could be expressed a bitt more succinctly, doing the matching and the query in a single loop. Maybe something like this?

```
CompoundObjectPtr result;
ScenePath currentPath;
for( const auto &p : queryPath )
{
    currentPath.push_back( p );
    if( !result && currentPath.size() > refPath.size() || p != refPath[currentPath.size()-1] )
    {
       // We're past the common root
        result = new CompoundObject;
    }
    if( result )
    {
       // Query attributes and update result
    }
}
```

johnhaddon commented on src/GafferScene/BranchCreator.cpp:
```diff
@@ -131,6 +131,52 @@ void validateDestination( const ScenePlug::ScenePath &destination )
+	curPath.resize( matchingLength + 1 );
+	for( unsigned int i = matchingLength; i < queryPath.size(); i++ )
+	{
+		IECore::ConstCompoundObjectPtr curAttributes = inPlug->attributes( curPath );
```
This makes a temporary context internally - might be worth reusing our own PathScope externally and using `attributesPlug()->getValue()` instead?

johnhaddon commented on startup/gui/menus.py:
```diff
@@ -543,6 +543,7 @@ def __usdLightCreator( lightType ) :
 
 nodeMenu.append( "/USD/Attributes", GafferUSD.USDAttributes, searchText = "USDAttributes" )
 nodeMenu.append( "/USD/Layer Writer", GafferUSD.USDLayerWriter, searchText = "USDLayerWriter" )
+nodeMenu.append( "/USD/Promote Instances", GafferUSD.PromoteInstances, searchText = "PromoteInstances" )
```
Following on from my earlier comment, should the node be called PromotePointInstances, or is that too verbose?

johnhaddon commented on python/GafferUSD/PromoteInstances.gfr:
```diff
@@ -0,0 +1,240 @@
+import Gaffer
```
I'm not seeing any unit tests for this node, but there definitely seems to be enough subtlety in the implementation that we should have a good few.

johnhaddon commented on python/GafferSceneUI/SceneViewUI.py:
```diff
@@ -651,20 +657,28 @@ def hasLabel( self ) :
 		m.append( "/Expand All Divider", { "divider" : True } )
 		m.append( "/Expand All", { "checkBox" : expandAll, "command" : Gaffer.WeakMethod( self.__toggleMinimumExpansionDepth ) } )
+		m.append( "/USD Instancers Divider", { "divider" : True } )
+		m.append( "/Expand USD Instancers When Rendering", { "checkBox" : expandUsdInstancers, "command" : Gaffer.WeakMethod( self.__toggleExpandUsdInstancers ) } )
```
This feels kindof wordy. I guess you're trying to get across the difference between using OpenGL and a renderer like Arnold? Is it worth having a separate option for each - seems like it might be useful to expand in OpenGL sometimes (but it would default off)?

danieldresser-ie replied:
Happy to do whatever you think is best. Something like a top level menu "Expand USD Instancers" with child checkboxes menu options "OpenGL" and "CPU Renderers" or something? Not sure what to call the latter option - "Not OpenGL" is ugly, but it's probably unnecessary to have separate checkboxes for every renderer we support?

Note that adding an OpenGL option here makes the interaction I mentioned before between Hierarchy Expansion and locations that don't exist in the original scene even more of a problem - OpenGL doesn't support capsules, and in a production scene, you're usually not going to want to have "Expand All" on in OpenGL. I suspect users will be confused when they turn on "Expand USD Instancers > OpenGL", and don't see any instances, because "Expand All" isn't on.

danieldresser-ie replied:
As per conversation, this now shows in the UI as just "Expand USD Instancers", but there is a different value for each renderer. This is implemented as an AtomicCompoundData plug on the adaptor with a BoolData per renderer ... feels like a bit of a weird interface, but it's fully internal.

johnhaddon commented on startup/GafferUSD/usdInstancers.py:
```diff
@@ -0,0 +1,150 @@
+# into actual Instancers. This supports how USD PointInstancers are currently loaded from
+# USD files. In the future, they may be first class objects, and this will be unnecessary.
+
+def __usdPointInstancerAdaptor() :
```
I think we'd benefit from some unit tests for this node. I suggest we move it to `GafferUSD._PointInstancerAdaptor` (note the underscore for protection), add tests in `GafferUSDTest.PointInstancerAdaptorTest` and then keep the registration here as a one-liner.

danieldresser-ie replied:
Should it be stored as a .gfr as well, for consistency with the PromoteInstances node?

danieldresser-ie replied:
You didn't feel strongly about this ... I ended up deciding not to store it as a .gfr, because the rewiring of an Instancer necessary to make it recursive wouldn't be very clear that way.

johnhaddon commented on startup/GafferUSD/usdInstancers.py:
```diff
@@ -0,0 +1,150 @@
+	# to get all the paths, and a Collect to remove the ones where the attribute value is
+	# set to 0.
+	result["usdPointInstancerSet"] = GafferScene.SetFilter()
+	result["usdPointInstancerSet"]["setExpression"].setValue( 'usd:pointInstancers' )
```
I'm getting errors when rendering scenes in Arnold where I've already done my own manual Instancer+SetFilter. This is a concern, because it's likely to be what everyone else has been doing to date. Deleting the `usd:pointInstancers` set fixes it for me, but I don't know the exact mechanism for the error, which is this :

```
00:06:00  1814MB WARNING |  [ginstance] /MediterraneanHills/Buildings/instances1: trying to clone a NULL object
```

That's using this very simple node network :

<details>

```
import Gaffer
import GafferScene
import imath

Gaffer.Metadata.registerValue( parent, "serialiser:milestoneVersion", 1, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:majorVersion", 5, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:minorVersion", 2, persistent=False )
Gaffer.Metadata.registerValue( parent, "serialiser:patchVersion", 0, persistent=False )

__children = {}

__children["SceneReader1"] = GafferScene.SceneReader( "SceneReader1" )
parent.addChild( __children["SceneReader1"] )
__children["SceneReader1"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["Instancer"] = GafferScene.Instancer( "Instancer" )
parent.addChild( __children["Instancer"] )
__children["Instancer"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SetFilter"] = GafferScene.SetFilter( "SetFilter" )
parent.addChild( __children["SetFilter"] )
__children["SetFilter"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
__children["SceneReader1"]["fileName"].setValue( '/home/john/data/caches/usd/city/PointInstancedMedCity.usd' )
__children["SceneReader1"]["__uiPosition"].setValue( imath.V2f( 9.60000038, 6.0999999 ) )
__children["Instancer"]["in"].setInput( __children["SceneReader1"]["out"] )
__children["Instancer"]["filter"].setInput( __children["SetFilter"]["out"] )
__children["Instancer"]["prototypes"].setInput( __children["SceneReader1"]["out"] )
__children["Instancer"]["prototypeMode"].setValue( 1 )
__children["Instancer"]["prototypeIndex"].setValue( 'prototypeIndex' )
__children["Instancer"]["omitDuplicateIds"].setValue( False )
__children["Instancer"]["orientation"].setValue( 'orientation' )
__children["Instancer"]["scale"].setValue( 'scale' )
__children["Instancer"]["inactiveIds"].setValue( 'inactiveIds invisibleIds' )
__children["Instancer"]["attributePrefix"].setValue( 'user:' )
__children["Instancer"]["__capsuleScene"]["bound"].setInput( __children["Instancer"]["out"]["bound"] )
__children["Instancer"]["__capsuleScene"]["transform"].setInput( __children["Instancer"]["out"]["transform"] )
__children["Instancer"]["__capsuleScene"]["attributes"].setInput( __children["Instancer"]["out"]["attributes"] )
__children["Instancer"]["__capsuleScene"]["globals"].setInput( __children["Instancer"]["out"]["globals"] )
__children["Instancer"]["__capsuleScene"]["setNames"].setInput( __children["Instancer"]["out"]["setNames"] )
__children["Instancer"]["__uiPosition"].setValue( imath.V2f( 8.10000038, -4.76406384 ) )
__children["SetFilter"]["setExpression"].setValue( 'usd:pointInstancers' )
__children["SetFilter"]["__uiPosition"].setValue( imath.V2f( 25.7000008, 4.19999933 ) )


del __children
```

</details>

danieldresser-ie replied:
Looks Arnold outputs this warning when expanding a procedural with no contents, which currently happens whenever you have an encapsulated instancer filtered to locations with no geo. It should be straightforward to put an extra for whether inPlug()->objectPlug()->getValue() is a NullObject in Instancer::computeObject before making a capsule, which should fix the warning. But it won't fix the perhaps more confusing issue in this scenario: nothing rendering, because we filter out any locations that are children of the instancer set, and the user has applied the instancer, but hasn't removed the location from the set.

danieldresser-ie replied:
Per our conversation, we now filter out locations where there is no Primitive with a "P" variable as not counting as part of the pointInstancers set, and therefore their children are not removed.

johnhaddon commented on python/GafferUSD/PromoteInstances.gfr:
```diff
@@ -0,0 +1,240 @@
+__children["PromoteInstances"]["Expression4"]["__out"].addChild( Gaffer.StringPlug( "p1", direction = Gaffer.Plug.Direction.Out, defaultValue = '', flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
+__children["PromoteInstances"]["Expression4"].addChild( Gaffer.V2fPlug( "__uiPosition", defaultValue = imath.V2f( 0, 0 ), flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
+__children["PromoteInstances"].addChild( GafferScene.PrimitiveVariableTweaks( "PrimitiveVariableTweaks4" ) )
+__children["PromoteInstances"]["PrimitiveVariableTweaks4"]["tweaks"].addChild( Gaffer.TweakPlug( Gaffer.IntPlug( "value", defaultValue = 0, ), "tweak0", flags = Gaffer.Plug.Flags.Default | Gaffer.Plug.Flags.Dynamic, ) )
```
It's early days for our use of ExtensionAlgo, so it's understandable that we haven't settled on some standards for it yet. And admittedly I am also running out of steam at the end of the day, but I do think this could be a bit clearer in terms of the naming of the nodes and their layout. For example, is there a better name than PrimitiveVariableTweaks4, could we use dots so the connections don't obscure the backdrop text, and could we size the backdrops so the text doesn't overflow (or perhaps use annotations)?

I could imagine a "no backdrops, no dots, no annotations" stance on the grounds that they're not needed at runtime. But I went pretty much the opposite way with ContactSheet and tried to make the graph as neat and documented as I could. I think that's a good place to start - later on we can always make ExtensionAlgo strip such things on export as a sort of "compilation" step from "source".

johnhaddon commented on python/GafferUSD/PromoteInstances.gfr:
```diff
@@ -0,0 +1,240 @@
+Gaffer.Metadata.registerValue( parent["variables"]["imageCataloguePort"], 'readOnly', True )
+Gaffer.Metadata.registerValue( parent["variables"]["projectName"]["name"], 'readOnly', True )
+Gaffer.Metadata.registerValue( parent["variables"]["projectRootDirectory"]["name"], 'readOnly', True )
+Gaffer.Metadata.registerValue( __children["PromoteInstances"], 'description', 'Convert points in a point cloud into instanced geometry. Assumes the point cloud matches the conventions of a USD PointInstancer.' )
```
Would be good to go into a bit more detail here, describing how we deactivate points in the instancer etc.

johnhaddon reviewed (commented):
Thanks Daniel! It's nice to have the expansion working "out of the box", and the new PromoteInstances node seems like it'll be very useful in production.

I'd like to think I've done a reasonably thorough review of most of this, but I did run out of steam a bit on PromoteInstances. I'm hoping a quick layout/naming pass on your side might help me get in a bit deeper there when I come back to it.

> There are lots of names here that I'm not very confident in - happy to switch to whatever wording feels best to people.

I've made a few suggestions in that regard, but I'm not super confident in them. Happy to consider alternatives too.

Cheers...
John

johnhaddon commented:
One other thing in case I forget - we'll need to update Changes.md before we can merge.

danieldresser-ie commented on startup/GafferUSD/usdInstancers.py:
```diff
@@ -0,0 +1,150 @@
+
+	return result
+
+GafferScene.SceneAlgo.registerRenderAdaptor( "USDPointInstancerAdaptor", __usdPointInstancerAdaptor, "*", "*" )
```
I wasn't thinking of any other clients where this could apply - I guess the render pass editor can trigger adaptors? Makes sense that we want to avoid it applying there.

danieldresser-ie commented on python/GafferSceneUI/StandardAttributesUI.py:
```diff
@@ -279,6 +279,36 @@ def __motionBlurSummary( plug ) :
 
+		# USD Instancers
+
+		"attributes.expandUsdInstancers" : [
+
+			"description",
+			"""
+			Controls whether USD PointInstancers should be automatically converted to
+			Gaffer instances at render time, so the render will include all final
+			instances. USD PointInstancers are currently represented in Gaffer as
+			PointsPrimitives in the usd:pointInstancers set. Without this attribute,
+			the render will just show the point cloud and the prototype objects.
+			""",
+
+			"layout:section", "USD Instancers",
+
+		],
+
+		"attributes.usdInstancerAttributes" : [
+
+			"description",
+			"""
+			Names of primitive variables that should be converted to attributes with
+			the user: prefix on instances automatically created from USD PointInstancer.
+			These attributes can then be used to drive shaders that differ between instances.
+			""",
+
+			"layout:section", "USD Instancers",
+
+		],
```
I spoke about it with Murray, we concluded StandardAttributes was more appropriate than USDAttributes - I guess the USDAttributes description currently makes this pretty clear: `Authors attributes which have specific meaning in USD, but which do not influence Gaffer's native behaviour in any way (in which case they would belong on the StandardAttributes node).` - though we could change that if you thought it was appropriate. But it sounds like you're fine with just not exposing these attributes anywhere in the UI?

danieldresser-ie commented on src/GafferSceneUI/SceneView.cpp:
```diff
@@ -1996,12 +1998,32 @@ SceneView::SceneView( ScriptNodePtr scriptNode )
+	PathFilterPtr rootsFilter = new PathFilter( "rootsFilter" );
+	preprocessor->addChild( rootsFilter );
+	rootsFilter->pathsPlug()->setValue( new StringVectorData( { "/*" } ) );
+	expandUsdInstancers->filterPlug()->setInput( rootsFilter->outPlug() );
```
I was able to get this working with some cheeky hacking - maybe you can take a look and suggest how this should actually be done.

It took 3 commits:
01916785e22d6aa6594b9b71df32803f3157f252 - I really wasn't sure what you meant by "reuse popupMenuSignal". We could just actually reuse to PlugValueWidget.popupMenuSignal(), and just always check if the type of the widget is _ExpansionPlugValueWidget in our callback - oh, thinking about it now, that probably is what you meant? I didn't like the idea of adding an extra callback that always needs to be called centrally ... but this only happens when you trigger a menu, that would probably be fine actually.

50452458f79bb5b5270bf6d48dd358a8a4c269c3 - I put a name on the RenderAdaptors so it's easier to find them when introspecting ... maybe there should be some sort of accessor here instead?

8503a6f6ce306ce4dd7385193cb150027f8141af - Actually add the feature - not sure whether this direct access of `__preprocessor` is too naughty, but it's pretty simple at least.

danieldresser-ie replied:
This has now been cleaned up per our conversation: the specific menu signal is now just named _ExpansionPlugValueWidget.menuSignal()

johnhaddon commented on startup/GafferSceneUI/usdPointInstancerAdaptor.py:
```diff
@@ -0,0 +1,54 @@
+def __expandUSDPointInstancersMenu( menuDefinition, plugValueWidget ) :
+
+	sceneView = plugValueWidget.getPlug().node()
+	renderAdaptor = sceneView["__preprocessor"]["RenderAdaptors"]["USDPointInstancerAdaptor"]["SceneProcessor"]
```
It seems likely that Cinesite will deregister this adaptor in favour of their own, with more bells and whistles. So we'll need to turn this into a no-op if we don't find our own adaptor.

danieldresser-ie commented:
OK, I think all feedback has now been addressed or replied to.

I think everything should be in a pretty good place at this point.

There were the two prerequisite changes that helped simplify the network for PromotePointInstances - I ended up doing both. `invertSelection` in PrimitiveVariableTweaks wasn't really necessary, but it was extremely easy to add. Support for casting data in TweakPlug is perhaps a bit trickier, but I realized it would otherwise be extremely cumbersome to support all possible types of input inactiveIds masks in the network.

You had a concern about inconsistency of types being output by PrimitiveVariableTweaks when using converted types ... I don't recall it exactly? I don't think it's that bad? Most modes never change the input type, and simply fail if the input isn't found. Create always outputs the tweak type. The only mode that can output either a source type or the tweak type is CreateIfMissing mode ... but I haven't changed that at all here.

The other annoying thing about the new TweakPlug code is that my initial version failed to compile on MSVC. I haven't quite figured out whether it was out of spec ... the spec says "Outside a template, a discarded statement is fully checked." ... and it's not clear to me whether a lambda using auto counts as a template ... I guess perhaps not. But if that's the case, and MSVC was going by the spec, it should have failed trying to compile the beginning of the lambda, as soon as I reference TweakDataType::ValueType, since it gets compiled with a data type of Data* ( if it is actually not discarding constexpr if, then it shouldn't discard on the IsTypedData check ). Instead, that seems to succeed, but then it fails when I do `tweakConverted.reserve`, compiling it for non-vector types despite the IsVectorTypedData check. The fact it succeeds in a simple case but fails in a complex one has me feeling like it's more likely to be a compiler bug, but we still need to work around it somehow. I guess that last commit isn't too horrid.

danieldresser-ie commented:
OK, I've cleaned up that fix to TweaksPlug that got things working on MSVC a bit, and addressed those last few comments about PromotePointInstances.

Question that might be worth asking before we merge this: if we're planning on switching to loading PointInstancers as having relative prototype paths ( controlled with an env var for backwards compatibility ), will we need to make PromotePointInstances and _PointInstancesAdaptor check that env var for whether to set the relativePrototypeRoots flag on the Instancer node? Or do we release these new nodes as something that only works if you have the env var set to relative mode?

danieldresser-ie commented:
Forgot to mention - I've got no idea what's going on with the Windows test failures now. Previously, I was dealing with some template compilation failure that we're clearly caused by me. But these new failures really don't seem related at all to what I'm doing, and didn't happen before I rebased ontop of other recent changes. Anyone else got any idea what's happening here?

danieldresser-ie commented:
> will we need to make PromotePointInstances and _PointInstancesAdaptor check that env var for whether to set the relativePrototypeRoots flag on the Instancer node?

Replying to myself here: there's probably no reason not to just make these nodes check the env var. I wasn't sure how we would implement checking the env var so it doesn't happen repeatedly ( in C++ we usually use global statics ) ... but if we just have an expression with no inputs that reads the env var, that should be cached like anything else ... that's probably fine.

johnhaddon commented:
> will we need to make PromotePointInstances and _PointInstancesAdaptor check that env var for whether to set the relativePrototypeRoots flag on the Instancer node?

I hadn't anticipated there being a `relativePrototypeRoots` flag on the instancer node - that's a level of faff I don't want to expose users to. I was expecting paths starting with "/" to be treated as absolute and others to be treated as relative. I suppose there is some concern about a change of behaviour there for old paths without a leading "/", but I still hope we can do this without another plug.

> I've got no idea what's going on with the Windows test failures now

I think that is unrelated to this PR. We'll find out soon, because I've merged!
